### PR TITLE
In file-encryptor sample, removed padding after decryption.

### LINKED ISSUE #3851: The file-encryptor sample has a wrong decryption practice
Making use of CBC mode, the decryption fails to remove the padding at the end of the recovered plaintext, which actually makes the output not equal to the original file.

```
[~/openenclave-0.14.0/samples/file-encryptor/build]$ hexdump -C ../testfile | tail
00000290  66 69 72 73 74 20 67 6f  6f 64 0a 67 72 65 61 74  |first good.great|
000002a0  65 72 20 63 61 6c 6c 65  64 20 74 6f 67 65 74 68  |er called togeth|
000002b0  65 72 20 64 6f 6d 69 6e  69 6f 6e 2e 20 55 6e 74  |er dominion. Unt|
000002c0  6f 20 77 69 74 68 6f 75  74 20 73 75 62 64 75 65  |o without subdue|
000002d0  20 63 72 65 65 70 69 6e  67 20 76 6f 69 64 20 79  | creeping void y|
000002e0  6f 75 0a 64 6f 6d 69 6e  69 6f 6e 20 74 77 6f 20  |ou.dominion two |
000002f0  6d 61 6e 20 66 6f 72 74  68 20 6f 6e 65 20 6d 6f  |man forth one mo|
00000300  76 69 6e 67 20 61 62 75  6e 64 61 6e 74 6c 79 20  |ving abundantly |
00000310  6d 75 6c 74 69 70 6c 79  2e 0a                    |multiply..|
0000031a
[~/openenclave-0.14.0/samples/file-encryptor/build]$ hexdump -C out.decrypted | tail
00000290  66 69 72 73 74 20 67 6f  6f 64 0a 67 72 65 61 74  |first good.great|
000002a0  65 72 20 63 61 6c 6c 65  64 20 74 6f 67 65 74 68  |er called togeth|
000002b0  65 72 20 64 6f 6d 69 6e  69 6f 6e 2e 20 55 6e 74  |er dominion. Unt|
000002c0  6f 20 77 69 74 68 6f 75  74 20 73 75 62 64 75 65  |o without subdue|
000002d0  20 63 72 65 65 70 69 6e  67 20 76 6f 69 64 20 79  | creeping void y|
000002e0  6f 75 0a 64 6f 6d 69 6e  69 6f 6e 20 74 77 6f 20  |ou.dominion two |
000002f0  6d 61 6e 20 66 6f 72 74  68 20 6f 6e 65 20 6d 6f  |man forth one mo|
00000300  76 69 6e 67 20 61 62 75  6e 64 61 6e 74 6c 79 20  |ving abundantly |
00000310  6d 75 6c 74 69 70 6c 79  2e 0a 06 06 06 06 06 06  |multiply........|
00000320
```

### PR DESCRIPTION
**Pull request by RRathna**
Fix #3851

### CODE REVIEW

File: samples/file-encryptor/host/host.cpp [hunk 1]
```diff
@@ -266,16 +269,22 @@ int encrypt_file(
             goto exit;
         }
 
+        bytes_to_write = bytes_read;
```
Comment (mingweishih) [inline samples/file-encryptor/host/host.cpp:272]: Can you elaborate a bit on what the bug is? It seems that original logic around `bytes_left` already handles the last block. If the `file_data_size` is less than the `DATA_BLOCK_SIZE`, doesn't the `bytes_read` returned by `fread` equals to `file_data_size`?

Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:272]: The size of testfile is 794 bytes and that of out.encrypted file is 800 bytes (with 6 bytes of padding).  When encrypting, we add  6 bytes of padding bytes to the testfile, to get block size that works out for CBC. 

In previous implementation when decrypting, we were taking in 800 bytes of out.encrypted and decrypting it to get 800 bytes of out.decrypted. The last 6 bytes were the padding bytes in this file. We shouldn't be writing the padding bytes to the out.decrypted file. In this change, I am writing only the first 794 bytes (hence the need for data_left).

bytes_left -> initialized to size of data in out.encrypted
data_left -> initialized to size of data in testfile (from header.file_data_size)

Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:272]: Checkout the last few bytes in testfile and out.decrypted files in before (1st execution) and after (2nd execution) the change is applied.
out decrypted results

Comment (mingweishih) [reply samples/file-encryptor/host/host.cpp:272]: It seems to me that the bug might just be here
https://github.com/openenclave/openenclave/blob/f1f9635e0c2a00d69210ab7f259b00052124cf52/samples/file-encryptor/host/host.cpp#L190
Would change this to
```
src_data_size = header.file_data_size; 
```
help?

Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:272]: So the last block of out.encrypted file needs to be passed in with the extra bytes added for padding to encrypt_block() (as these are needed to retrieve the initial data)...

File: samples/file-encryptor/host/host.cpp [hunk 2]
```diff
@@ -266,9 +267,17 @@ int encrypt_file(
...
+        // During decryption, check for and remove padding from last block.
```
[the remark this answers is not in the document]
Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:271]: No, the last block is only expected to be a multiple of CIPHER_BLOCK_SIZE, whereas if bytes_left is less than or equal to DATA_BLOCK_SIZE, it is the last block. In this implementation, we have

#define CIPHER_BLOCK_SIZE 16
#define DATA_BLOCK_SIZE 256

Eg. data size in testfile= 500 bytes. The out.encrypted file will be 512 bytes.
when decrypting, the second block read from out.encrypted will be 256 bytes( ie; bytes_left == DATA_BLOCK_SIZE) . After decrpytion, we need to write only the first 244 bytes after decryption onto out.decrypted.

File: samples/file-encryptor/host/host.cpp [hunk 3]
```diff
@@ -266,9 +267,17 @@ int encrypt_file(
...
+        if (!encrypt && bytes_left <= DATA_BLOCK_SIZE)
+        {
+            bytes_to_write = header.file_data_size % DATA_BLOCK_SIZE;
```
[the remark this answers is not in the document]
Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:274]: The 'if' block  will be executed only once during a file decryption (when the bytes_left is less than or equal to DATA_BLOCK_SIZE). There will be a check at every block processing. I guess the cost of variable increment + check would be more or same?

File: samples/file-encryptor/host/host.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply samples/file-encryptor/host/host.cpp:271]: Ah, the sentence seems wrong. Let me rephrase
```
        // The data size is always padded to align with DATA_BLOCK_SIZE after encryption.
        // Therefore, removing the padding (if any) from the last block after decryption
        // and write the last block to the output file.
```
Does it match what you describe?

File: samples/file-encryptor/host/host.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply samples/file-encryptor/host/host.cpp:274]: Thanks for the clarification.

File: samples/file-encryptor/host/host.cpp [hunk 4]
```diff
@@ -266,9 +267,17 @@ int encrypt_file(
             goto exit;
         }
 
+        bytes_to_write = bytes_read;
+        // During decryption, check for and remove padding from last block.
+        if (!encrypt && bytes_left <= DATA_BLOCK_SIZE)
```
[the remark this answers is not in the document]
Comment (mingweishih) [reply samples/file-encryptor/host/host.cpp:272]: The padding is CIPHER_BLOCK_SIZE- instead of DATA_BLOCK_SIZE-alignment, so both cases can happen.

File: samples/file-encryptor/host/host.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (RRathna) [reply samples/file-encryptor/host/host.cpp:274]: yeah, this sounds better!

**Review by mingweishih — APPROVED** [review]
LGTM. Thanks for fixing this.

File: samples/file-encryptor/host/host.cpp [hunk 5]
```diff
@@ -266,9 +270,18 @@ int encrypt_file(
             goto exit;
         }
 
+        bytes_to_write = bytes_read;
+        // The data size is always padded to align with DATA_BLOCK_SIZE during encryption.
+        // Therefore, removing the padding (if any) from the last block during decryption.
```
Comment (dthaler) [inline samples/file-encryptor/host/host.cpp:275]:
```suggestion
        // Therefore, remove the padding (if any) from the last block during decryption.
```

**Review by dthaler — APPROVED** [review]
Looks ok other than grammar nit

File: samples/file-encryptor/host/host.cpp (hunk 5 above)
Comment (RRathna) [reply samples/file-encryptor/host/host.cpp]: Implemented.

**Comment by RRathna** [comment]
discarding this PR as commits conflict with another branch's commits (have no idea why). Creating a new PR.
